### DISCUSSION TITLE
Questions about the working of file plugin

### DISCUSSION THREAD

**Discussion by ssmntn210** [coredns/coredns#5329] [Q&A] [answered]
Hi everyone,

I was working with CoreDNS and had a couple of questions about the same. I was exploring the file plugin for a use case and was wondering how exactly it parses and indexes the zone file, which would help us deciding if we are to go ahead with it. 

On going through the codebase, I see that the zone file is parsed and stored as a left leaning red black tree, and the queries are a essentially search traversal on this LLRB tree. Hence, I expect the latency to be in order of O(log n) wrt to number of entries, but on experimenting, I see that the graph of latency vs number of records in the zone file is not logarithmic but rather constant, hovering around a same value. Follow up on the same : 
1) I assume that the LLRB tree is constructed at the start when the coredns is first loaded (or rather whenever its reloaded), and everytime a query is made, the plugin is triggered and the same LLRB instance is used to service the query. Is the assumption correct? If yes, is the tree already sitting in memory?
2) What indexing is used on the zone file to make the queries? Is it not LLRB? If it is, then why the flat curve, if not then how is it indexed?
3) Is there any caching middle layer that causes this latency to be somewhat constant, or could it be other latencies that overshadow the O(log n) factor?

Thanks a lot, any help would be greatly appreciated!

**Comment by chrisohaver** [comment, marked as answer]
The zone file is not re-parsed for every query, the tree stays in memory.  There is no caching built into the _file_ plugin (I am assuming you tested without enabling the _cache_ plugin).

Can you share details about your testing method in a such a manner that it can be reproduced by others?  And also share your testing results?

**Reply by ssmntn210** [reply]
Thanks for the answer, that clears my doubt about the tree. 

I had the cache enabled, I ensured that the queries serviced are not in cache by registering and issuing new queries. Did not think the direction of disabling the cache plugin 😅, would probably try it once.

The following is the description of the experiment:
Append 400 entries to the file and re-deploy coredns (this ensured that the new zone file - which is loaded through hostPath and volumeMount - is correctly loaded and that the cache is cleared for sure). Queries are made _sequentially_  from foo0.bar to foo(n).bar, where n is the total number of entries registered in the zone file (hence ensuring each of them are not serviced from the cache). The latencies are stored and at the end I record the average latency. The process is repeated till I have 48,000 entries in the zone file. Couple of points:
 1) The number of queries (and hence the denominator in the average) is increasing, I tried another setup where instead of querying n domains, I query only the last 400 domains registered. Got a similar result: constant time throughout (with slightly more variations).
 2) I guess restarting the coredns pod instead of re-deploying would work too, was doing it just too be double sure. 
 3) The script is written in Go, and is run in the same VM that hosts the cluster
 
 The following is the result:
 
<img width="580" alt="image" src="https://user-images.githubusercontent.com/99073970/164749719-f84c485c-fdd3-4cc0-b216-04c2ecf23462.png">

I would try the experiment once again after disabling the cache plugin too, but other than that, any catch here?

**Reply by chrisohaver** [reply]
It's possible that the overhead of sending - receiving - processing - replying - receiving the response  is dwarfing the tree search time.  IOW, that everything else other than searching the tree is taking much more time than the time it takes to search the tree.  If so, then the tree search time may be insignificant and not noticeable in comparison.  The flat line suggests this may be the case (although there is a very slight rise I think?).

Am I correct in seeing that he lower number of entry tests have less data points - and hence a spikier "average" for the small N tests?  It would be better to test with the same number of queries for each test  - e.g. 100k+ queries each test, the more the better (disabling cache plugin would help here). 

Another more complex approach could be to measure just tree search time, by adding tooling into the code to expose a metric via prometheus, and building your own test build of coredns.

**Reply by ssmntn210** [reply]
Yes, the explanation makes sense. True, there is a slight rise after 40,000 mark and spikes before the 10,000. Let me experiment by making a good, constant number of queries at every data point with cache disabled.

One last question, if the trees stays in the memory, what causes the difference between the latencies of queries served using _cache_ plugin to be faster than the ones serviced from the _file_ plugin, which is the case as I noticed in an another set of experiments? Is due to the use of a more suitable data structure or hardware optimisations?

**Comment by chrisohaver** [comment]
> True, there is a slight rise after ...

I meant to say I suspect there is a slight slope upwards in the whole set, if you were to best fit a line to it.  The rise after 40K could system noise (are you doing this on a dedicated system, or a workstation/laptop?).

> what causes the difference between the latencies of queries served using cache plugin to be faster than the ones serviced from the file plugin

I'm not sure that's being observed in your data, since you are not using cache (each record only queried once), records are never served from cache.

The primary use case for the _cache_ plugin is to reduce the number of queries to upstream nameservers (with the _forward_ plugin), or for plugins that have expensive record lookups (maybe a plugin with a database backend for example).  The file plugin does not have a very expensive record lookup, so it doesn't really benefit all that much when used in conjunction with the cache plugin.

**Reply by ssmntn210** [reply]
> I meant to say I suspect there is a slight slope 


I guess the slope would be very slight, but agreed..
> I'm not sure that's being observed in your data, since you are not using cache  ....


I carried a different set of experiments where I query the same domain 2000 times, and then 2000 different domains and compare them. Turned out that the average latency for uncached scenario was 1.8 times more than that of cached scenario, although this is in order of tenths of millisecond.
